### PR TITLE
[EigenApp] EigenSystemSolver: optional selection of MKL

### PR DESCRIPTION
**Pull request by philbucher**
Now it is possible to select the linear solver used inside the `EigensystemSolver`
this does not compile as is but I think it shows what I want to do
Maybe someone has an idea how to make it work :)

this way it can be ensure e.g. in tests that no matter if MKL is enabled or not we get the same results

@manuelmessmer this should help you deal with your issues

### CODE REVIEW

**Comment by philbucher** [comment]
hm as I thought, selecting the type at runtime does not work :)
We could use polymorphism to solve this. 
However since the `EigenPardisoLDLTSolver` and the `EigenSparseLUSolver` do not have a common baseclass I think we have two options:
- Introduce a common baseclass. Larger change but then we could use more different solvers too
- make small wrapper classes like explained [here](https://stackoverflow.com/a/11197069) that we could use for this

Are there alternative other solutions?

**Comment by armingeiser** [comment]
I think a small wrapper class inside the eigensystem solver is enough. Adding a base class for all eigensolvers just for the sake of using it here is overkill IMO.

However i think this is only necessary if we do not find the reason for the different results in #5938 
Because if the behaviour would be as expected, why would one not use MKL if available.

**Comment by philbucher** [comment]
I like being able to select the solver. It might not be necessary often, but sometimes.

**Comment by philbucher** [comment]
@oberbichler your templates gave me quite a hard time :sweat_smile: 
However I think this is because of `Eigen`

File: applications/EigenSolversApplication/custom_solvers/eigensystem_solver.h [hunk 1]
```diff
@@ -307,6 +313,34 @@ class EigensystemSolver
...
+
+        virtual ~DirectSolverWrapperBase() = default;
+        virtual void Compute(MatrixMapType a) = 0;
+        virtual void Solve(ConstVectorRefType b, VectorRefType x) = 0;
+    };
+
+    template<class TSolver>
+    struct DirectSolverWrapper : DirectSolverWrapperBase
+    {
+        typedef DirectSolverWrapper BaseType;
```
Comment (armingeiser) [inline applications/EigenSolversApplication/custom_solvers/eigensystem_solver.h]:
```suggestion
        typedef DirectSolverWrapperBase BaseType;
```

**Review by armingeiser — APPROVED** [review]
For me it looks good.

**Pull request merged**